File: services/python/sanic/src/application.py
import asyncio
from sanic import Sanic
from sanic.exceptions import NotFound, MethodNotAllowed, PayloadTooLarge

from app.state import AppState
from routes.hooks import log_hook, chaos_hook
from routes.handlers import (
    post_handler,
    get_handler,
    delete_handler,
    put_handler,
    patch_handler,
    internal_handler,
    invalid_method_handler,
    invalid_path_handler,
    large_payload_handler,
)

# TODO: make configurable?
MAX_BODY_SIZE: int = 64 * 1024


class Application:
    def __init__(self) -> None:
        self.app = Sanic("sanic", configure_logging=False)
        self.app.ctx.state = AppState()
        self.app.ctx.tasks_lock = asyncio.Lock()
        self.app.config.REQUEST_MAX_SIZE = MAX_BODY_SIZE

        self.app.on_request(chaos_hook)
        self.app.on_request(log_hook)

        self.app.error_handler.add(NotFound, invalid_path_handler)
        self.app.error_handler.add(MethodNotAllowed, invalid_method_handler)
        self.app.error_handler.add(Exception, internal_handler)
        self.app.exception(PayloadTooLarge)(large_payload_handler)

        self.app.post("/")(post_handler)
        self.app.get("/<id:uuid>")(get_handler)
        self.app.delete("/<id:uuid>")(delete_handler)
        self.app.put("/<id:uuid>")(put_handler)
        self.app.patch("/<id:uuid>")(patch_handler)

File: services/go/echo/application.go
package main

import (
	"app"
	"echo/routes/handlers"
	"echo/routes/middleware"

	"github.com/labstack/echo/v5"
	echomw "github.com/labstack/echo/v5/middleware"
)

const maxBodySize = 64 * 1024

type Application struct {
	Engine *echo.Echo
	state  *app.AppState
}

func NewApplication() *Application {
	state := app.NewState()

	engine := echo.NewWithConfig(echo.Config{
		Router: echo.NewRouter(echo.RouterConfig{
			NotFoundHandler:         handlers.InvalidPathHandler,
			MethodNotAllowedHandler: handlers.InvalidMethodHandler,
		}),
	})
	engine.Use(echomw.BodyLimit(maxBodySize))
	engine.Use(middleware.RecoverMiddleware)
	engine.Use(middleware.LogMiddleware)
	engine.Use(middleware.ChaosMiddleware)

	engine.POST("/", handlers.PostHandler(state))
	engine.GET("/:id", handlers.GetHandler(state))
	engine.DELETE("/:id", handlers.DeleteHandler(state))
	engine.PATCH("/:id", handlers.PatchHandler(state))
	engine.PUT("/:id", handlers.PutHandler(state))

	return &Application{Engine: engine, state: state}
}
